exporter/elasticsearchexporter/metric_grouping.go

```diff
@@ -69,14 +102,23 @@ func (h otelDataPointHasher) hashDataPoint(resource pcommon.Resource, scope pcom
 
 	hasher.Write([]byte(dp.Metric().Unit()))
 
-	mapHashExcludeReservedAttrs(hasher, dp.Attributes(), elasticsearch.MappingHintsAttrKey)
+	mapHashSortedExcludeReservedAttrs(hasher, dp.Attributes(), elasticsearch.MappingHintsAttrKey)
 
 	return hasher.Sum32()
 }
 
-// mapHashExcludeReservedAttrs is mapHash but ignoring some reserved attributes.
+func (h otelDataPointHasher) hashCombined(_ pcommon.Resource, _ pcommon.InstrumentationScope, _ datapoints.DataPoint) uint32 {
+	return 0
+}
+
+// mapHashSortedExcludeReservedAttrs is mapHash but ignoring some reserved attributes and is independent of order in Map.
 // e.g. index is already considered during routing and DS attributes do not need to be considered in hashing
-func mapHashExcludeReservedAttrs(hasher hash.Hash, m pcommon.Map, extra ...string) {
+func mapHashSortedExcludeReservedAttrs(hasher hash.Hash, m pcommon.Map, extra ...string) {
```

Comment:
Did we consider using common libraries like:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/pdatautil/hash.go
https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/internal/exp/metrics/identity

for the hash calculations? I think almost all the functions here could be replaced by existing hash options. For the exclusion of `DataStream*` attributes we could create a custom hash option.

Reply:
Probably not able to use identity directly, since some fields are used but are not TSDB dimensions, e.g. schema url.

pdatautil hash, on the other hand, looks good and will get rid of a lot of the hashing code we have. Still need to look into how to achieve exclusions. I think we'll be better off creating a map and passing it in, instead of writing an Option which means reimplementing all the funcs in pdatautil hash because they are private.

Reply:
> I think we'll be better off creating a map and passing it in

which will involve a fair bit of copying

Reply:
> Probably not able to use identity directly, since some fields are used but are not TSDB dimensions, e.g. schema url.

I thought so, the identity is a bit rigid.

> Still need to look into how to achieve exclusions

We can implement a new [HashOption](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9bb671ca7ab3a9e0861d068f15b146aaa55766c2/pkg/pdatautil/hash.go#L35) (in elasticsearch-exporter for now) - something like `WithMapExclusions(m pcommon.Map, excludedKeys ...string) HashOption`

Reply:
> instead of writing an Option which means reimplementing all the funcs in pdatautil hash because they are private

Ah, right!!! I didn't see the comment. Maybe, add a TODO for exporter and create a PR in pdatautil to have this function?

Reply:
Added a TODO, shouldn't be a blocker for this PR.